app/src/main/java/com/example/habitsmasher/UserValidator.java

```diff
@@ -0,0 +1,153 @@
+    private static final String PASSWORD_LENGTH_MUST_BE_GREATER_THAN_6_MESSAGE = "Password length must be greater than 6!";
+    private static final String PASSWORD_LENGTH_MUST_BE_LESS_THAN_16_MESSAGE = "Password length must be less than 16!";
+
+    private EditText _usernameInput;
+    private EditText _emailInput;
+    private EditText _passwordInput;
+
+    String _username;
+    String _email;
+    String _password;
+
+    // for testing
+    public UserValidator() {}
+
+    // for sign up
+    public UserValidator(EditText usernameInput,
+                         EditText emailInput,
+                         EditText passwordInput) {
+        _usernameInput = usernameInput;
+        _emailInput = emailInput;
+        _passwordInput = passwordInput;
+
+        // save text contents from EditText fields
+        _email = _emailInput.getText().toString().trim();
+        _password = _passwordInput.getText().toString().trim();
+        _username = _usernameInput.getText().toString().trim();
+    }
+
+    // for log in
+    public UserValidator(EditText emailInput,
+                         EditText passwordInput) {
+        _usernameInput = null;
+        _emailInput = emailInput;
+        _passwordInput = passwordInput;
+
+        // save text contents from EditText fields
+        _email = _emailInput.getText().toString().trim();
+        _password = _passwordInput.getText().toString().trim();
+        _username = "";  // since there is no username for logging in, empty
+    }
+
+    public String getValidUsernameForSignUp() {
+        if (isNewUserValid()) {
+            return _username;
+        }
+        return "";
+    }
+    public String getValidEmailForSignUp() {
+        if (isNewUserValid()) {
+            return _email;
+        }
+        return "";
+    }
+    public String getValidPasswordForSignUp() {
+        if (isNewUserValid()) {
+            return _password;
+        }
+        return "";
+    }
+
+    public String getValidEmailForLogin() {
+        if (isLoginValid()) {
+            return _email;
+        }
+        return "";
+    }
+    public String getValidPasswordForLogin() {
+        if (isLoginValid()) {
+            return _password;
+        }
+        return "";
+    }
+
+    private boolean isNewUserValid() {
+        return isNewUserValid(_username, _email, _password);
+    }
+
+    private boolean isLoginValid() {
+        return isLoginValid(_email, _password);
+    }
+
+    public boolean isNewUserValid(String username, String email, String password) {
+        return !(username.isEmpty() | username.length() > 15 | !isLoginValid(email, password));
+    }
+
+    public boolean isLoginValid(String email, String password) {
+        return !(email.isEmpty() | password.isEmpty() |
+                !PatternsCompat.EMAIL_ADDRESS.matcher(email).matches() |
+                password.length() < 6 |
+                password.length() > 15);
+    }
```

Comment:
What was the rationale for splitting up the isVaild into two methods here?

Reply:
I was more asking about why there are two variants of each isVaild method. There are two isNewUserValid methods and two isLoginValid methods.

Reply:
I see. I'm just **overriding** these methods to use internally. You can see they are `private`. They are used in the getters/setters to keep those methods short (since we already have access to username, email, and password at the class level, no need to pass it in). 

These methods with no parameters end up calling the main methods anyways. This is a common design pattern.